openexr-rs/src/tiledinputfile.rs

```diff
@@ -0,0 +1,970 @@
+    ///
+    /// Return value is a Box2i (imath) with the min value
+    /// `(data_window.min.x + dx * TiledInputFile::tile_x_size(), data_window.min.y + dy * TiledInputFile::tile_y_size())`
+    /// and max value
+    /// `(data_window.min.x + (dx + 1) * TiledInputFile::tile_x_size() - 1, data_window.min.y + (dy + 1) * TiledInputFile::tile_y_size() - 1)`
+    ///
+    pub fn data_window_for_tile<T: Bound2<i32>>(
+        &self,
+        dx: i32,
+        dy: i32,
+        lx: i32,
+        ly: i32,
+    ) -> Result<T, Error> {
+        let mut data_window =
+            std::mem::MaybeUninit::<sys::Imath_Box2i_t>::uninit();
+
+        unsafe {
+            sys::Imf_TiledInputFile_dataWindowForTile(
+                self.inner,
+                data_window.as_mut_ptr(),
+                dx,
+                dy,
+                lx,
+                ly,
+            )
+            .into_result()?;
+
+            let data_window: [i32; 4] =
+                std::mem::transmute(data_window.assume_init());
+            Ok(T::from_slice(&data_window))
+        }
+    }
+
+    /// Reads the tile with tile coordinates `(dx, dy)`, and level number
+    /// `(lx, ly)`, and stores it in the current frame buffer.
+    ///
+    /// # Errors
+    ///
+    /// Read tile will return an error if the following rules are not followed:
+    ///
+    /// - dx must lie in the interval `[0, TiledInputFile::num_x_tiles(lx)-1]`
+    /// - dy must lie in the interval `[0, TiledInputFile::num_y_tiles(ly)-1]`
+    /// - lx must lie in the interval `[0, TiledInputFile::num_x_levels()-1]`
+    /// - ly must lie in the interval `[0, TiledInputFile::num_y_levels()()-1]`
+    ///
+    pub fn read_tile(
+        &mut self,
+        dx: i32,
+        dy: i32,
+        lx: i32,
+        ly: i32,
+    ) -> Result<(), Error> {
+        unsafe {
+            sys::Imf_TiledInputFile_readTile(self.inner, dx, dy, lx, ly)
+                .into_result()?;
+        }
+
+        Ok(())
+    }
+
+    /// This allows reading multiple tiles at once. If multi-threading is used
+    /// the multiple tiles are read concurrently.
+    ///
+    /// Pixels that are outside the pixel coordinate range for the
+    /// tile's level, are never accessed by the read tile.
+    ///
+    /// # Errors
+    ///
+    /// Attempting to access a tile that is not present in the file
+    /// returns an InputExc error.
+    ///
+    pub fn read_tiles(
+        &mut self,
+        dx1: i32,
+        dx2: i32,
+        dy1: i32,
+        dy2: i32,
+        lx: i32,
+        ly: i32,
+    ) -> Result<(), Error> {
+        unsafe {
+            sys::Imf_TiledInputFile_readTiles(
+                self.inner, dx1, dx2, dy1, dy2, lx, ly,
+            )
+            .into_result()?;
+        }
+
+        Ok(())
+    }
+
+    /// Read a tile of raw pixel data from the file, without uncompressing it
+    /// (this function is used to implement TiledOutputFile::copy_pixels()).
```

Comment:
Do we want to expose raw tile data? From the sounds of the comment here it's not really a public API function. Would probably be cleaner just to ignore these methods for now and come back if we need them. Reading uncompressed data directly is more in OpenEXRCore's domain, once we get to that.